fe2-android/app/src/test/java/com/github/dedis/student20_pop/model/data/LAORepositoryTest.java

```diff
@@ -37,121 +43,201 @@
+    testScheduler.advanceTimeTo(RESPONSE_DELAY, TimeUnit.MILLISECONDS);
+    testObserver.assertComplete().assertNoErrors()
+        .assertValue(r -> r.getId() == REQUEST_ID);
+  }
+
+  @Test
+  public void testGetAllLaos() {
+    // Does not use LAOLocalDataSource
+
+    // Create the LAO and the subscriber to test from
+    Lao lao = new Lao(LAO_CHANNEL);
+    TestObserver<List<Lao>> subscriber = TestObserver.create();
 
-    testScheduler.advanceTimeBy(950, TimeUnit.MILLISECONDS);
-    testObserverUnsubscribe.assertNotComplete();
+    // Subscribe to a LAO and wait for the request to finish
+    repository.sendSubscribe(LAO_CHANNEL);
+    testScheduler.advanceTimeBy(RESPONSE_DELAY, TimeUnit.MILLISECONDS);
 
-    testScheduler.advanceTimeBy(100, TimeUnit.MILLISECONDS);
-    testObserverUnsubscribe.assertComplete().assertNoErrors()
-        .assertValue(r -> r.getId() == 42);
+    repository.getAllLaos().subscribe(subscriber);
+
+    // Check the LAO is present in the laoById map of LAORepository
+    List<Lao> repositoryLaoList = (List<Lao>) subscriber.getEvents().get(0).get(0);
+    assertEquals(repositoryLaoList.get(0).getChannel(), lao.getChannel());
   }
 
-  // Test the get methods: getLaoObservable and getAllLaos
   @Test
-  public void testGetLaos() {
+  public void testGetLaoObservable() {
     // Does not use LAOLocalDataSource
 
-    // Send subscribe to fill laoById and allLaoSubject in LAORepository
-    repository.sendSubscribe("/root/123");
-    TestObserver<Lao> testObserverLao = TestObserver.create();
-    TestObserver<List<Lao>> testObserverListLao = TestObserver.create();
+    // Subscribe to a LAO and wait for the request to finish
+    repository.sendSubscribe(LAO_CHANNEL);
+    testScheduler.advanceTimeBy(RESPONSE_DELAY, TimeUnit.MILLISECONDS);
+
+    // Set subscriber of the LAO list to test from
+    TestObserver<Lao> subscriber = TestObserver.create();
+    repository.getLaoObservable(LAO_CHANNEL).subscribe(subscriber);
 
-    testScheduler.advanceTimeBy(1500, TimeUnit.MILLISECONDS);
+    // Check the LAO is present in the allLaoSubject list of LAORepository
+    Lao repositoryLao = (Lao) subscriber.getEvents().get(0).get(0);
+    assertEquals(repositoryLao.getChannel(), LAO_CHANNEL);
+  }
+
+  @Test
+  public void testBroadcast() throws GeneralSecurityException {
+    // Destroy the instance from the setup
+    LAORepository.destroyInstance();
+
+    // Create message to broadcast
+    byte[] dataBuf = Injection.provideGson().toJson(CREATE_LAO, Data.class).getBytes();
+    Mockito.when(signer.sign(dataBuf)).thenReturn(dataBuf);
+    MessageGeneral message = new MessageGeneral(Base64.getUrlDecoder().decode(ORGANIZER), CREATE_LAO, signer, Injection.provideGson());
+
+    // Simulate a network response and then a broadcast from the server after the response delay
+    Observable<GenericMessage> upstream = Observable.fromArray((GenericMessage) new Result(REQUEST_ID), (GenericMessage) new Broadcast(LAO_CHANNEL, message))
```

Comment:
```suggestion
    Observable<GenericMessage> upstream = Observable.fromArray(
        (GenericMessage) new Result(REQUEST_ID),
        (GenericMessage) new Broadcast(LAO_CHANNEL, message))
```
might be more readable